src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -191,61 +198,84 @@ protected Solution SetCurrentSolution(Solution solution)
             Func<Solution, Solution> transformation,
             WorkspaceChangeKind kind,
             ProjectId? projectId = null,
-            DocumentId? documentId = null)
+            DocumentId? documentId = null,
+            Action<Solution, Solution>? onBeforeUpdate = null,
+            Action<Solution, Solution>? onAfterUpdate = null)
```

Comment:
helper that handles the common case of issuing the WorkspaceChangedEvent.  Just defers to the full-flexibility helper below it.